### PR TITLE
EnC: Handle anonymous types and delegates hoisted to closure or state machine

### LINKED ISSUE #3233: Assert in CSharpSymbolMatcher.MatchSymbols.TryFindAnonymousType
%mymachine%\dumps\AnonType20150601.dmp
%mymachine%\dumps\AnonType20150601.zip

```C#
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ConsoleApplication139
{
    class Program
    {
        static void Main(string[] args)
        {
            var result =
                from a in args
                let x = a.Reverse()
                let y = x.Reverse()
                where x.SequenceEqual(y)
                select new { Value = a, Length = a.Length };
            var first = result.First();
            args = args.Concat(from r in result select r.Value.ToString()).ToArray();
            result.ToString();
        }
    }
}
```

1. Do a bunch of stuff.
1. Run to ```result.ToString()```.
1. Change ```a.Length``` to ```a.Length + 1```.
1. Set next statement to ```var result = ...``` (might actually happen before previous step)
1. Step.

### PR DESCRIPTION
**Pull request by tmat**
Fixes 1170899: fatal error during EnC.
Fixes #3233, #3192.

**Scenarios**
When editing a method multiple times that contains a variable of an anonymous type (or delegate in VB) that is hoisted to a closure or a state machine the debugger reports a fatal error and terminates debugging session. 

Examples:
```VB
    Sub Main()
        Dim x = Function(n As Integer) n + 1
        Dim y = Sub(n As Integer) Console.WriteLine(x(n))
        y(100)
    End Sub
```

```C# 
        static void Main(string[] args)
        {
            var result =
                from a in args
                let x = a.Reverse()
                let y = x.Reverse()
                where x.SequenceEqual(y)
                select new { Value = a, Length = a.Length };            
        }
```

**Fix**
We didn't deeply translate anonymous types/delegates before comparing them in symbol matcher.
The proposed fix for RTM isn't ideal but addresses the issues and avoids risky changes. A proper fix requires extensive changes in symbol translations and should be implemented post RTM.

**Risk**
Although not ideal I'm confident that that fix is addressing the root cause of the issue. The fix introduces new code paths which are all covered by newly added unit tests. These code paths are only exercised during EnC specific code generation and emit and are not taken during regular compilation, which limits the impact only to EnC scenarios.

**Testing**
Added unit tests and manually verified that affected scenarios work for both VB and C#. Buddy tested by @wschae and @amcasey. 

@ManishJayaswal

### CODE REVIEW

**Comment by tmat** [comment]
@cston @AlekseyTs @wschae @amcasey

File: src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs [hunk 1]
```diff
@@ -1316,7 +1316,8 @@ private BoundStatement ChainImplicitStructConstructor(MethodSymbol methodSymbol,
                 if (optimizations == OptimizationLevel.Debug && stateMachineTypeOpt != null)
                 {
                     Debug.Assert(method.IsAsync || method.IsIterator);
-                    GetStateMachineSlotDebugInfo(moduleBuilder.GetSynthesizedFields(stateMachineTypeOpt), out stateMachineHoistedLocalSlots, out stateMachineAwaiterSlots);
+                    GetStateMachineSlotDebugInfo(moduleBuilder, moduleBuilder.GetSynthesizedFields(stateMachineTypeOpt), diagnostics, out stateMachineHoistedLocalSlots, out stateMachineAwaiterSlots);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs]: Should this be 'diagnosticsForThisMethod'? Or can we remove 'diagnosticsForThisMethod' and use 'diagnostics' everywhere?

File: src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs [hunk 2]
```diff
@@ -752,11 +751,31 @@ public SymbolComparer(MatchSymbols matcher)
...
+                        left = AnonymousTypeManager.TranslateAnonymousTypeSymbol(anonymousLeft);
+                    }
+
+                    var anonymousRight = right as AnonymousTypeManager.AnonymousTypePublicSymbol;
+                    if ((object)anonymousRight != null)
+                    {
+                        right = AnonymousTypeManager.TranslateAnonymousTypeSymbol(anonymousRight);
+                    }
+
+                    if ((object)anonymousLeft != null || (object)anonymousRight != null)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs]: Would this get any simpler extracting a helper method?
if (TranslateIfAnonymousTypePublicSymbol(ref left) || TranslateIfAnonymousTypePublicSymbol(ref right)) { ... }

File: src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodToStateMachineRewriter.cs [hunk 3]
```diff
@@ -96,7 +96,7 @@ private FieldSymbol GetAwaiterField(TypeSymbol awaiterType)
             if (!_awaiterFields.TryGetValue(awaiterType, out result))
             {
                 int slotIndex;
-                if (slotAllocatorOpt == null || !slotAllocatorOpt.TryGetPreviousAwaiterSlotIndex((Cci.ITypeReference)awaiterType, out slotIndex))
+                if (slotAllocatorOpt == null || !slotAllocatorOpt.TryGetPreviousAwaiterSlotIndex(F.ModuleBuilderOpt.Translate(awaiterType, null, F.Diagnostics), out slotIndex))
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodToStateMachineRewriter.cs]: Is there an assert that could be added to TryGetPreviousAwaiterSlotIndex and TryGetPreviousHoistedLocalSlotIndex to catch untranslated types?

Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodToStateMachineRewriter.cs]: Pass F.Syntax for syntax argument.

File: src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs [hunk 4]
```diff
@@ -752,11 +751,31 @@ public SymbolComparer(MatchSymbols matcher)
...
-                    var other = _matcher.Visit(x);
-                    Debug.Assert((object)other != null);
-                    return other == y;
+                    if (left.Kind != right.Kind)
+                    {
+                        return false;
+                    }
+
+                    var anonymousLeft = left as AnonymousTypeManager.AnonymousTypePublicSymbol;
+                    if ((object)anonymousLeft != null)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs]: if (left.IsAnonymousType)

File: src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/PublicSymbols/AnonymousType.TypePublicSymbol.cs [hunk 5]
```diff
@@ -22,7 +22,7 @@ internal sealed partial class AnonymousTypeManager
         /// Represents an anonymous type 'public' symbol which is used in binding and lowering.
         /// In emit phase it is being substituted with implementation symbol.
         /// </summary>
-        private sealed class AnonymousTypePublicSymbol : NamedTypeSymbol
+        internal sealed class AnonymousTypePublicSymbol : NamedTypeSymbol
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/PublicSymbols/AnonymousType.TypePublicSymbol.cs]: Does this need to be internal if SymbolMatcher uses .IsAnonymousType?

File: src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs (hunk 4 above)
Comment (tmat) [inline src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs]: That would require left to be TypeSymbol, which it is not.

File: src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/PublicSymbols/AnonymousType.TypePublicSymbol.cs (hunk 5 above)
Comment (tmat) [inline src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/PublicSymbols/AnonymousType.TypePublicSymbol.cs]: Perhaps not, but the code would be more complex. Also the corresponding type is already internal in VB.

File: src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodToStateMachineRewriter.cs (hunk 3 above)
Comment (tmat) [inline src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodToStateMachineRewriter.cs]: Re assert: for most types translated types are the same instance. We would need to check for specific translated types (like NoPIA, anonymous, etc.) But we can't currently do that easily in language agnostic way :(

Comment (tmat) [inline src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodToStateMachineRewriter.cs]: F.Syntax ... Done.

Comment (tmat) [inline src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodToStateMachineRewriter.cs]: Post RTM we could probably introduce struct TranslatedSymbol<T> and have the type system to check it for us.

File: src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs (hunk 1 above)
Comment (tmat) [inline src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs]: Yes.

File: src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/SynthesizedSymbols/AnonymousType.PropertySymbol.cs [hunk 6]
```diff
@@ -19,7 +19,7 @@ internal sealed partial class AnonymousTypeManager
         /// <summary>
         /// Represents an anonymous type template's property symbol.
         /// </summary>
-        private sealed class AnonymousTypePropertySymbol : PropertySymbol
+        internal sealed class AnonymousTypePropertySymbol : PropertySymbol
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/SynthesizedSymbols/AnonymousType.PropertySymbol.cs]: Is it necessary to make AnonymousTypePropertySymbol, AnonymousTypeTemplateSymbol, and NameAndIndex internal?

Comment (tmat) [inline src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/SynthesizedSymbols/AnonymousType.PropertySymbol.cs]: Once AnonymousTypePublicSymbol is the other classes need to as well, since they are used by non-private members of AnonymousTypePublicSymbol.

File: src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicDefinitionMap.vb [hunk 7]
```diff
@@ -121,10 +123,10 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Emit
                         Case GeneratedNameKind.StateMachineAwaiterField
 
                             If GeneratedNames.TryParseSlotIndex(StringConstants.StateMachineAwaiterFieldPrefix, name, slotIndex) Then
-                                Dim field = TryCast(member, IFieldSymbol)
+                                Dim field = DirectCast(member, IFieldSymbol)
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicDefinitionMap.vb]: Can we use FieldSymbol rather than IFieldSymbol?

**Comment by cston** [comment]
LGTM

File: src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs [hunk 8]
```diff
@@ -653,7 +653,6 @@ private static MethodSymbol SubstituteTypeParameters(MethodSymbol method)
 
             private bool AreNamedTypesEqual(NamedTypeSymbol type, NamedTypeSymbol other)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs]: I find the name of this method very confusing. It looks like its logic is specific to Anonymous Types, perhaps the name should reflect this and asserting on the input night be appropriate.

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueClosureTests.cs [hunk 9]
```diff
@@ -2625,5 +2625,148 @@ .maxstack  3
...
+  IL_0021:  call       ""void System.Console.WriteLine(int)""
+  IL_0026:  nop
+  IL_0027:  ret
+}");
+
+            var diff2 = compilation2.EmitDifference(diff1.NextGeneration,
+                ImmutableArray.Create(
+                    new SemanticEdit(SemanticEditKind.Update, f1, f2, GetSyntaxMapFromMarkers(source1, source2), preserveLocalVariables: true)));
+
+            diff1.VerifySynthesizedMembers(
```
Comment (wschae) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueClosureTests.cs]: diff2?

File: src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs (hunk 2 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs]: Should we be using ```&&``` instead?

File: src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicSymbolMatcher.vb [hunk 10]
```diff
@@ -510,7 +510,6 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Emit
             End Function
 
             Private Function AreNamedTypesEqual(type As NamedTypeSymbol, other As NamedTypeSymbol) As Boolean
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicSymbolMatcher.vb]: Comments for C# version of this function is applicable here as well. In addition, it doesn't look like this function makes any distinction between Anonymous Types and Anonymous Delegates. It feel like the distinction should be made.

File: src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs (hunk 8 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs]: Names of the members are significant for Anonymous Types. Should we compare them?

File: src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicSymbolMatcher.vb [hunk 11]
```diff
@@ -587,8 +586,26 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Emit
...
+                    If anonymousLeft IsNot Nothing Then
+                        left = anonymousLeft.MapToImplementationSymbol()
+                    End If
+
+                    Dim anonymousRight = TryCast(right, AnonymousTypeManager.AnonymousTypeOrDelegatePublicSymbol)
+                    If anonymousRight IsNot Nothing Then
+                        right = anonymousRight.MapToImplementationSymbol()
+                    End If
+
+                    If anonymousLeft IsNot Nothing OrElse anonymousRight IsNot Nothing Then
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicSymbolMatcher.vb]: AndAlso?

File: src/Compilers/VisualBasic/Test/Emit/Emit/EditAndContinue/EditAndContinueTestBase.vb [hunk 12]
```diff
@@ -21,6 +22,13 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.UnitTests
 
         Friend Shared ReadOnly EmptyLocalsProvider As Func(Of MethodDefinitionHandle, EditAndContinueMethodDebugInformation) = Function(token) Nothing
 
+        Friend Shared Function Visualize(baseline As ModuleMetadata, ParamArray deltas As PinnedMetadata()) As String
```
Comment (wschae) [inline src/Compilers/VisualBasic/Test/Emit/Emit/EditAndContinue/EditAndContinueTestBase.vb]: Is it for future reference? I don't see anyone calling it.

File: src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs (hunk 2 above)
Comment (tmat) [inline src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs]: No. One of the symbols may be public symbol the other may be PE symbol of the implementation.

File: src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs (hunk 8 above)
Comment (tmat) [inline src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs]: The method is not specific to anon types.

Yes, good point.

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueClosureTests.cs (hunk 9 above)
Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueClosureTests.cs]: Yes. Thanks.

File: src/Compilers/VisualBasic/Test/Emit/Emit/EditAndContinue/EditAndContinueTestBase.vb (hunk 12 above)
Comment (tmat) [inline src/Compilers/VisualBasic/Test/Emit/Emit/EditAndContinue/EditAndContinueTestBase.vb]: Yes, it's a helper one can use to diagnose issues.

**Comment by tmat** [comment]
@cston @AlekseyTs I have revisited the implementation of comparer in SymbolMatcher, hydration of PE symbols from initial metadata in PEDeltaAssemblyBuilder and added more tests. Please, review the updated change.

File: src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs [hunk 13]
```diff
@@ -752,11 +759,41 @@ public SymbolComparer(MatchSymbols matcher)
...
+                    }
+
+                    if ((object)anonymousLeft != null || (object)anonymousRight != null)
+                    {
+                        var namedLeft = (NamedTypeSymbol)left;
+                        var namedRight = (NamedTypeSymbol)right;
+
+                        // Map the template - this involves comparing the structure of the anonymous type,
+                        // so we don't compare equal types with different field names.
+                        if (_matcher.Visit(namedLeft.ConstructedFrom)?.Equals(namedRight.ConstructedFrom) != true)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs]: Would == be sufficient? if (_matcher.Visit(namedLeft.ConstructedFrom) == namedRight.ConstructedFrom) { ... }

File: src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs [hunk 14]
```diff
@@ -1316,7 +1316,8 @@ private BoundStatement ChainImplicitStructConstructor(MethodSymbol methodSymbol,
                 if (optimizations == OptimizationLevel.Debug && stateMachineTypeOpt != null)
                 {
                     Debug.Assert(method.IsAsync || method.IsIterator);
-                    GetStateMachineSlotDebugInfo(moduleBuilder.GetSynthesizedFields(stateMachineTypeOpt), out stateMachineHoistedLocalSlots, out stateMachineAwaiterSlots);
+                    GetStateMachineSlotDebugInfo(moduleBuilder, moduleBuilder.GetSynthesizedFields(stateMachineTypeOpt), diagnosticsForThisMethod, out stateMachineHoistedLocalSlots, out stateMachineAwaiterSlots);
+                    Debug.Assert(!diagnostics.HasAnyErrors());
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs]: diagnosticsForThisMethod

File: src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs (hunk 13 above)
Comment (tmat) [inline src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs]: Yes.

**Comment by cston** [comment]
LGTM

File: src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs [hunk 15]
```diff
@@ -343,6 +349,7 @@ public override Symbol Visit(Symbol symbol)
 
                 // Add an entry for the match, even if there is no match, to avoid
                 // matching the same symbol unsuccessfully multiple times.
+                var x = base.Visit(symbol);
```
Comment (wschae) [inline src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs]: Is it for debugging purpose?

**Comment by MattGertz** [comment]
MoreInfo 6/1 by ML Shiproom.  Please get buddy-testing on this.  It clearly meets the bar, but we just want to make sure that we have no regressions because this is a big change touching a lot of files.

File: src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs (hunk 13 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs]: Is it possible to get into a cycle here (ConstructedFrom can return the same symbol back)?

File: src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/SynthesizedSymbols/AnonymousType.TemplateSymbol.cs [hunk 16]
```diff
@@ -133,7 +133,7 @@ internal AnonymousTypeTemplateSymbol(AnonymousTypeManager manager, AnonymousType
 
             internal AnonymousTypeKey GetAnonymousTypeKey()
             {
-                var properties = this.Properties.SelectAsArray(p => AnonymousTypeKeyField.CreateField(p.Name));
+                var properties = this.Properties.SelectAsArray(p => new AnonymousTypeKeyField(p.Name, isKey: false, ignoreCase: false));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/SynthesizedSymbols/AnonymousType.TemplateSymbol.cs]: What is the reason for this change? Please revert unless behavior has changed.

File: src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs (hunk 15 above)
Comment (tmat) [inline src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs]: Yes, forgot to update the PR. I have removed it.

File: src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/SynthesizedSymbols/AnonymousType.TemplateSymbol.cs (hunk 16 above)
Comment (tmat) [inline src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/SynthesizedSymbols/AnonymousType.TemplateSymbol.cs]: AnonymousTypeKeyField did convert the name to lower-case (when it was created from VB) instead of comparing case-insensitively. Lowering case and then comparing ordinal is incorrect.

**Comment by AlekseyTs** [comment]
LGTM

**Comment by dpoeschl** [comment]
retest this please

**Comment by tmat** [comment]
@AlekseyTs @cston @amcasey @wschae @ManishJayaswal Figured out a fairly low risk workaround for the anonymous type mapping issues. Implemented as a separate commit on top of the previous PR. All issues that I know of are now addressed. Please, review as soon as tests pass.

**Comment by AlekseyTs** [comment]
All this magic around deep translation can benefit from detailed comments about why and what are we doing.

**Comment by tmat** [comment]
@AlekseyTs I'll add a comment to make it more clear.

**Comment by cston** [comment]
LGTM

**Comment by tmat** [comment]
@AlekseyTs I have removed the direct cats to ITypeReference and added the comment.

**Comment by AlekseyTs** [comment]
LGTM

**Comment by wschae** [comment]
Buddy Tested: 

- verified EnC not working w/o PR
- verified EnC working with PR
- intensively covered a combination usage of anonymous types, lambdas, async and iterators
- checked both C# and VB

LGTM

**Comment by amcasey** [comment]
I'm out of buddy testing ideas and it's been more than an hour since I broke anything EnC related. :)

:+1:

**Comment by pharring** [comment]
Approved 6/5 ML Shiproom.
@tmat Please add an issue to do the "right fix" to the compiler in Milestone 1.1

**Comment by tmat** [comment]
Filed workitem #3341
